src/geom/stroke.rs

```diff
@@ -48,13 +87,219 @@ impl PartialStroke<Abs> {
+                sep = ", ";
+            }
+            if let Smart::Custom(join) = &line_join {
+                write!(f, "{}join: {:?}", sep, join)?;
+                sep = ", ";
+            }
+            if let Smart::Custom(dash) = &dash_pattern {
+                write!(f, "{}dash: {:?}", sep, dash)?;
+                sep = ", ";
             }
-            (Smart::Custom(paint), Smart::Auto) => paint.fmt(f),
-            (Smart::Auto, Smart::Custom(thickness)) => thickness.fmt(f),
-            (Smart::Auto, Smart::Auto) => f.pad("<stroke>"),
+            if let Smart::Custom(miter_limit) = &miter_limit {
+                write!(f, "{}miter-limit: {:?}", sep, miter_limit)?;
+            }
+            write!(f, ")")?;
+            Ok(())
+        }
+    }
+}
+
+/// The line cap of a stroke
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub enum LineCap {
+    Butt,
+    Round,
+    Square,
+}
+
+impl Debug for LineCap {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            LineCap::Butt => write!(f, "\"butt\""),
+            LineCap::Round => write!(f, "\"round\""),
+            LineCap::Square => write!(f, "\"square\""),
+        }
+    }
+}
+
+/// The line join of a stroke
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub enum LineJoin {
+    Miter,
+    Round,
+    Bevel,
+}
+
+impl Debug for LineJoin {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            LineJoin::Miter => write!(f, "\"miter\""),
+            LineJoin::Round => write!(f, "\"round\""),
+            LineJoin::Bevel => write!(f, "\"bevel\""),
+        }
+    }
+}
+
+/// A line dash pattern
+#[derive(Clone, Eq, PartialEq, Hash)]
+pub struct DashPattern<T = Length, DT = DashLength<T>> {
+    /// The dash array.
+    pub array: Vec<DT>,
+    /// The dash phase.
+    pub phase: T,
+}
+
+impl<T: Debug, DT: Debug> Debug for DashPattern<T, DT> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "(array: (")?;
+        for (i, elem) in self.array.iter().enumerate() {
+            if i == 0 {
+                write!(f, "{:?}", elem)?;
+            } else {
+                write!(f, ", {:?}", elem)?;
+            }
+        }
+        write!(f, "), phase: {:?})", self.phase)?;
+        Ok(())
+    }
+}
+
+impl<T: Default> From<Vec<DashLength<T>>> for DashPattern<T> {
+    fn from(array: Vec<DashLength<T>>) -> Self {
+        Self { array, phase: T::default() }
+    }
+}
+
+/// The length of a dash in a line dash pattern
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum DashLength<T = Length> {
+    LineWidth,
+    Length(T),
+}
+
+impl From<Abs> for DashLength {
+    fn from(l: Abs) -> Self {
+        DashLength::Length(l.into())
+    }
+}
+
+impl<T> DashLength<T> {
+    fn finish(self, line_width: T) -> T {
+        match self {
+            Self::LineWidth => line_width,
+            Self::Length(l) => l,
+        }
+    }
+}
+
+cast_from_value! {
+    LineCap: "line cap",
```

Comment:
you could also derive `Cast` for these, then you get this for free. only works for string-like enums though.

Reply:
That's pretty neat, I will do that